File: apps/api/app/schemas/portal.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict

UserRole = Literal["admin", "operator", "viewer"]


class LoginInput(BaseModel):
    account: str
    password: str


class ChangePasswordInput(BaseModel):
    currentPassword: str
    newPassword: str


class UserCreateInput(BaseModel):
    name: str
    account: str
    password: str
    role: UserRole = "operator"
    appPermissions: list[str] | None = None


class UserUpdateInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = None
    account: str | None = None
    password: str | None = None
    role: UserRole | None = None
    enabled: bool | None = None
    appPermissions: list[str] | None = None


class EnterAppInput(BaseModel):
    confirmedConflict: bool = False

File: apps/web/src/shared/types/portal.ts
export type UserRole = "admin" | "operator" | "viewer";

export type AppStatus = "running" | "available" | "maintenance" | "offline";

export type ModuleCode =
  | "competitor-analysis"
  | "rag-web-search"
  | "contract-review"
  | "bid-generator"
  | "patent-disclosure";

export interface PortalUser {
  id: string;
  name: string;
  account: string;
  role: UserRole;
  enabled: boolean;
  appPermissions: ModuleCode[];
  createdAt?: string;
  updatedAt?: string;
  lastLoginAt?: string;
}

export interface PortalModule {
  code: ModuleCode;
  slug: "competitor-analysis" | "rag" | "contract-review" | "bid-generator" | "patent-disclosure";
  name: string;
  shortName: string;
  description: string;
  route: string;
  iframeRoute?: string;
  apiPrefix: string;
  legacyFrontend?: string;
  status: AppStatus;
}

export interface RuntimeAppConfig {
  code: ModuleCode;
  name: string;
  routePath?: string;
  frontendUrl?: string;
  backendUrl?: string;
  iframeUrl: string;
  url?: string;
  healthUrl?: string;
  enabled: boolean;
  devMode?: string;
}

export interface AppUsageSession {
  id: string;
  appId: ModuleCode;
  clientId: string;
  userId: string;
  userName: string;
  startedAt: string;
  lastActiveAt: string;
  confirmedConflict: boolean;
}

export interface AppUsageSummary {
  appId: ModuleCode;
  sessions: AppUsageSession[];
  otherUserSessions: AppUsageSession[];
  currentUserSessions: AppUsageSession[];
  inUse: boolean;
  inUseByOthers: boolean;
  userNames: string[];
  otherUserNames: string[];
}

export interface CreatePortalUserInput {
  name: string;
  account: string;
  password: string;
  role?: UserRole;
  appPermissions?: ModuleCode[];
}

export type UpdatePortalUserInput = Partial<
  Pick<PortalUser, "name" | "account" | "role" | "enabled" | "appPermissions">
> & {
  password?: string;
};

export interface ChangePasswordInput {
  currentPassword: string;
  newPassword: string;
}
